crates/runtime/README.md

```diff
@@ -0,0 +1,592 @@
+    D --> J[Storage Operations]
+    D --> K[Logging/Events]
+    D --> L[External Requests]
+    F --> N[VMLimits]
+```
+
+### VMLogic
+
+The `VMLogic` struct is the central component of the runtime. It encapsulates:
+
+  - Storage access
+  - Memory management
+  - Context information
+  - Resource limits
+  - Logging and event emission
+
+*[Read more...](#vmlogic-1)*
+
+### Host functions
+
+The runtime provides a set of host functions that can be called from the guest
+Wasm module:
+
+  - Memory and register operations
+  - Storage read/write
+  - Logging and event emission
+  - Cryptographic operations (signing and verification)
+  - External HTTP requests (fetch)
+
+*[Read more...](#host-functions-1)*
+
+### Resource limits
+
+The `VMLimits` struct defines constraints on various resources:
+
+  - Memory usage
+  - Stack size
+  - Number and size of registers
+  - Logging and event limits
+  - Storage key and value size limits
+
+### Storage
+
+The runtime includes an abstract `Storage` trait and an in-memory implementation
+(`InMemoryStorage`) for persisting state.
+
+*[Read more...](#storage-1)*
+
+### Error handling
+
+A comprehensive error handling system is implemented, covering various
+scenarios:
+
+  - Host errors
+  - Wasm execution errors
+  - Storage errors
+  - Function call errors
+
+*[Read more...](#error-handling-1)*
+
+## Operational flow
+
+  1. Load Wasm module
+  2. Set up execution environment (memory, limits, storage)
+  3. Resolve and link host functions
+  4. Execute specified function in the Wasm module
+  5. Handle return values, logs, and events
+  6. Clean up and return execution outcome
+
+```mermaid
+graph TD
+    A[Load Wasm Module] --> B[Set up Execution Environment]
+    B --> C[Create Wasmer Engine and Store]
+    C --> D[Set Memory Limits and Tuning]
+    D --> E[Instantiate Wasm Module]
+    E --> F[Resolve and Link Host Functions]
+    F --> G[Execute Specified Function]
+    G --> H{Execution Successful?}
+    H -->|Yes| I[Collect Results]
+    H -->|No| J[Handle Errors]
+    I --> K[Process Logs and Events]
+    J --> K
+    K --> L[Clean Up Resources]
+    L --> M[Return Execution Outcome]
+
+    subgraph "Host Function Calls"
+        G --> N[Memory Operations]
+        G --> O[Register Operations]
+        G --> P[Storage Interactions]
+        G --> Q[Logging and Events]
+        G --> R[External Requests]
+    end
+```
+
+## Detailed component descriptions
+
+### Topology
+
+```mermaid
+graph TD
+    B[VMLogic]
+    C[Wasm Module - Guest Code]
+    
+    B --> E[Storage]
+    B --> F[Memory Management]
+    B --> G[Registers]
+    B --> H[Context]
+    B --> I[Host Functions]
+    
+    I --> K[Memory Operations]
+    I --> L[Register Operations]
+    I --> M[Storage Operations]
+    I --> N[Logging/Events]
+    I --> O[External Requests]
+    
+    C -.-> I
+    
+    subgraph "External Environment"
+        AB[Host System]
+    end
+    
+    AB -.-> E
+    AB -.-> O
```

Comment:
wdyt about flipping these? since `Storage` depends on stuff provided by the host, and so does the request, being fulfilled by the host